blog/_posts/2021-05-26-wip-vs-harness.md

```diff
@@ -0,0 +1,138 @@
+---
+slug: "wip-vs-harness"
+date: 2021-05-26
+author: nick
+layout: blog
+title: "WIP Development vs Harness Development"
+subtitle: "Local K8s dev, GitOps, and when to use them together"
+description: "Local K8s dev, GitOps, and when to use them together"
+image: "/assets/images/wip-vs-harness/cover.jpg"
+image_caption: "<i>Everybody Rides in the Whip at Midland Beach</i>. Via <a href='https://digitalcollections.nypl.org/items/510d47d9-c024-a3d9-e040-e00a18064a99'>The New York Public Library</a>."
+tags:
+  - gitops
+  - localdev
+---
+
+In the #tilt chat last week, `renaudguerin` asked:
+
+> I’m the new DevOps guy in a small startup, and migrating us away from broken
+> in-house tooling towards a more modern GitOps workflow (ArgoCD/CodeFresh), and
+> local k8s dev environments (either Skaffold or garden.io, tilt.dev ...) 
+>
+> But how do you bridge the gap between the two ?
+
+This is a deep question! I've been thinking about an illuminating way to explore
+it.
+
+Here's an analogy that I've been toying with:
+
+Look around at some public Git repos you like. Pick some popular ones and some
+single-user ones.
+
+The single-user ones might have lots of commit messages that simply say "WIP".
+
+The popular ones have lots of instructions on how to review and test the PR.
+
+Imagine you've never used Git before. You look at the published commits. You
+conclude: these are mutually exclusive approaches. Some projects are WIP
+projects. And some projects are verbose projects.
+
+Maybe that analogy is illuminating. But if it's not, I'm going to go into more
+detail on some of the patterns we've seen on:
+
+- What problems GitOps & local Kubernetes dev solve
+
+- How these two toolsets approach those problems
+
+- How to bridge the gap between them
+
+## More Service, More Problems
+
+If you're working on a multi-service app, you have two types of changes:
+
+- WIP changes - No services are running in dev, and you want to make a change.
+
+- Harness changes - All your services are running in prod, and you want to make a change.
+
+These two sets may sound the same but they are mostly *non-overlapping*!
+
+Here's a table to help illustrate the differences:
+
+|          | WIP Changes | Harness Changes |
+----------------|-------------|------------------
+| Add segfaults to see what breaks                 | ✅ | ❌️                             |
+| Bring up all services from scratch               | ✅ | Something terrible has happened |
+| Reset the database                               | ✅️ | ❌ ❌ ❌                      |
+| Sacifice correctness for faster updates | ✅  |     ❌️                            |
+| Add security backdoors for testing | ✅ | 🚑 |
```

Comment:
Can we rephrase this to "security is not a concern" or something more neutral like that? "Add security backdoors for testing" has a very let's-kick-some-puppies feel to me.